aipmiddleware/validateparent.go

```diff
@@ -0,0 +1,89 @@
+package aipmiddleware
+
+import (
+	"context"
+
+	"go.einride.tech/aip/resourcename"
+	"google.golang.org/genproto/googleapis/api/annotations"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+)
+
+// ParentValidator validates that response resources are children of requested parent resources.
+type ParentValidator struct {
+	// LogHook is called when parent validation fails, with the failing request/response, parent/name combination.
+	LogHook func(info *grpc.UnaryServerInfo, request, response proto.Message, parent, name string)
+}
+
+// UnaryServerInterceptor implements grpc.UnaryServerInterceptor.
+func (p *ParentValidator) UnaryServerInterceptor(
+	ctx context.Context,
+	request interface{},
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler,
+) (interface{}, error) {
+	protoRequest, ok := request.(proto.Message)
+	if !ok {
+		return nil, status.Error(codes.Internal, "parent validator got non-proto request")
+	}
+	protoReflectRequest := protoRequest.ProtoReflect()
+	// Request must have a parent field of string kind.
+	parentField := protoReflectRequest.Descriptor().Fields().ByName("parent")
+	if parentField == nil ||
+		parentField.IsMap() ||
+		parentField.IsList() ||
+		parentField.Kind() != protoreflect.StringKind {
+		return handler(ctx, request)
+	}
+	// The parent field must have a resource reference with a child type.
+	parentResourceReference := proto.GetExtension(
+		parentField.Options(), annotations.E_ResourceReference,
+	).(*annotations.ResourceReference)
```

Comment:
We should handle `ok` here from the "casting", in case there is a field called parent but it does not have a resource reference.

Reply:
if it doesn't have a resource reference, we'll get nil here.

> GetExtension retrieves the value for an extension field. If the field is unpopulated, it returns the default value for scalars and an immutable, empty value for lists or messages. It panics if xt does not extend m.

https://pkg.go.dev/google.golang.org/protobuf/proto#GetExtension

Reply:
Amazing